src/dbSta/src/dbNetwork.cc

```diff
@@ -82,6 +85,77 @@ char* tmpStringCopy(const char* str)
   return tmp;
 }
 
+//
+// Handling of object ids (Hierachy Mode)
+//--------------------------------------
+//
+// The database assigns a number to each object. These numbers
+// are scoped based on the type. Eg dbModInst 1..N or dbInst 1..N.
+// The timer requires a unique id for each object for its visit
+// pattern, so we uniquify the numbers by suffixing a discriminating
+// address pattern to the lower bits and shifting.
+// Everytime a new type of timing related object is added, we
+// must update this code, so it is isolated and marked up here.
+//
+// The id is used by the STA traversers to accumulate visited.
+// lower 4  bits used to encode type
+//
+
+static constexpr unsigned DBITERM_ID = 0x0;
+static constexpr unsigned DBBTERM_ID = 0x1;
+static constexpr unsigned DBINST_ID = 0x2;
+static constexpr unsigned DBNET_ID = 0x3;
+static constexpr unsigned DBMODITERM_ID = 0x4;
+static constexpr unsigned DBMODBTERM_ID = 0x5;
+static constexpr unsigned DBMODINST_ID = 0x6;
+static constexpr unsigned DBMODNET_ID = 0x7;
+static constexpr unsigned DBMODULE_ID = 0x8;
+// Number of lower bits used
+static constexpr unsigned DBIDTAG_WIDTH = 0x4;
+
+ObjectId dbNetwork::getDbNwkObjectId(dbObjectType typ, ObjectId db_id) const
+{
```

Comment:
We should also probably have another check here to make sure the ids never go over what is allowed

```c++
if (db_id > (std::numeric_limits<ObjectId>::max() >> DBIDTAG_WIDTH)) {
  error
}
```